### PR TITLE
Allow pr reviewers to be set per maestro subscription

### PR DESCRIPTION
**Pull request by AdamYoblick**
Fixes https://github.com/dotnet/arcade/issues/1723

Starting with the darc client changes to make sure I'm on the right track, will tackle the back end changes afterwards

### CODE REVIEW

**Comment by AdamYoblick** [comment]
Hey @alexperovich these changes are done for the most part, but not tested yet. I need some guidance on how to add some new tables. Probably one for reviewers, and a many-to-many mapping table from reviewers to subscriptions

**Comment by alexperovich** [comment]
Since we use EF you can just add new properties here: https://github.com/dotnet/arcade-services/blob/master/src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs#L68 and classes to represent the tables. Probably a new class Reviewer and a Reviewers property. Then a SubscriptionReviewer class and property. Refer to the BuildChannels table on how to make that work.

Once you have changed the context code to match what you want you can then create a migration by running this in the Maestro.Data project directory:
```
dotnet ef --msbuildprojectextensionspath C:\path\to\repo\artifacts\obj\Maestro.Data migrations add <name>
```

That will update the BuildAssetRegistryContextModelSnapshot to match the changes you have made and add the class for the new migration.


After that is done you will need to update the API with the new endpoint, and add versioned model classes for the endpoint to use.

**Comment by AdamYoblick** [comment]
Ok thanks for the info Alex, I'll probably get back with you for guidance when I attempt to do the work. I just got pulled into something else, so this is going on hold for a little

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/UpdateReviewersOperation.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+            var reviewers = _options.Reviewers.Trim();
+
+            if (string.IsNullOrWhiteSpace(subscriptionId))
+            {
+                Logger.LogError("Please specify a subscription id");
+                return Constants.ErrorCode;
+            }
+
+            // stuff the reviewers into a List
+            var reviewersList = _options.Reviewers.Split(',').ToList();
```
Comment (mmitche) [inline src/Microsoft.DotNet.Darc/src/Darc/Operations/UpdateReviewersOperation.cs]: You can change the option itself to have a separator and then it returns an IEnumerable

https://github.com/dotnet/arcade-services/blob/master/src/Microsoft.DotNet.Darc/src/Darc/Options/AddSubscriptionCommandLineOptions.cs#L33

File: src/Microsoft.DotNet.Darc/src/Darc/Options/AddSubscriptionCommandLineOptions.cs [hunk 2]
```diff
@@ -43,6 +43,12 @@ class AddSubscriptionCommandLineOptions : CommandLineOptions
         [Option('q', "quiet", HelpText = "Non-interactive mode (requires all elements to be passed on the command line).")]
         public bool Quiet { get; set; }
 
+        /* todo
+        [Option("reviewers", HelpText = "Reviewers are automicatically added to all PRs created from this subscription. " + 
+            "Provide a comma separated list of user names.")]
+        public string Reviewers { get; set; }
```
Comment (mmitche) [inline src/Microsoft.DotNet.Darc/src/Darc/Options/AddSubscriptionCommandLineOptions.cs]: Why commented out here?

File: src/Microsoft.DotNet.Darc/src/Darc/Options/UpdateReviewersCommandLineOptions.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    internal class UpdateReviewersCommandLineOptions : CommandLineOptions
+    {
+        [Option("subscription-id", HelpText = "Id of the subscription to update.")]
+        public string SubscriptionId { get; set; }
+
+        [Option("reviewers", HelpText = "User names of the reviewers, comma separated. " +
+            "Note that this overwrites the entire reviewers list for this subscription")]
+        public string Reviewers { get; set; }
+
+        public UpdateReviewersCommandLineOptions()
```
Comment (mmitche) [inline src/Microsoft.DotNet.Darc/src/Darc/Options/UpdateReviewersCommandLineOptions.cs]: The constructor should be unnecessary

File: src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs [hunk 4]
```diff
@@ -295,6 +298,10 @@ public async Task<string> CreatePullRequestAsync(string repoUri, PullRequest pul
             };
             Octokit.PullRequest createdPullRequest = await Client.PullRequest.Create(owner, repo, pr);
 
+            // add reviewers to the created PR
+            PullRequestReviewRequest reviewers = new PullRequestReviewRequest(pullRequest.Reviewers.ToList());
+            await Client.PullRequest.ReviewRequest.Create(owner, repo, createdPullRequest.Number, reviewers);
```
Comment (mmitche) [inline src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs]: What happens here if an invalid name is passed? We probably should not fail to create the PR in that case

File: src/Maestro/Client/src/Generated/Models/SubscriptionUpdate.cs [hunk 5]
```diff
@@ -8,6 +8,7 @@
 namespace Microsoft.DotNet.Maestro.Client.Models
```
Comment (mmitche) [inline src/Maestro/Client/src/Generated/Models/SubscriptionUpdate.cs]: This file is generated off of the swagger API, it isn't manually edited.  So you add the right controller to SubscriptionsController, then do
```
npm install; npm start
```
From `src\Maestro\Client\generator`

File: src/Microsoft.DotNet.Darc/src/DarcLib/AzureDevOpsClient.cs [hunk 6]
```diff
@@ -286,13 +288,41 @@ public async Task<string> CreatePullRequestAsync(string repoUri, PullRequest pul
...
 
+            // get the identity client used to resolve user name strings to identities
+            var identityClient = await connection.GetClientAsync<IdentityHttpClient>();
+
+            // Get the identities of each reviewer
+            var identities = new List<IdentityRefWithVote>();
+            foreach (var reviewer in pullRequest.Reviewers)
+            {
+                // look for a matching display name
+                var matchingIdentity = (await identityClient.ReadIdentitiesAsync(IdentitySearchFilter.DisplayName, reviewer).ConfigureAwait(false)).FirstOrDefault();
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Darc/src/DarcLib/AzureDevOpsClient.cs]: >.ConfigureAwait(false) [](start = 124, length = 22)

Why are their configure awaits only on these methods> Three are other awaits in this method... in fact, there are other awaits this change added.

Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Darc/src/DarcLib/AzureDevOpsClient.cs]: There are a bunch of var's in this file that I don't know what they are referring to. Whta is "ReadIdentitiesAsync" retruning?  "var" is against our coding style.

File: src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs [hunk 7]
```diff
@@ -270,12 +270,15 @@ public async Task<PullRequest> GetPullRequestAsync(string pullRequestUrl)
...
+            var reviewers = pr.RequestedReviewers.Select(x => x.Name).ToList();
+
             return new PullRequest
             {
                 Title = pr.Title,
                 Description = pr.Body,
                 BaseBranch = pr.Base.Ref,
-                HeadBranch = pr.Head.Ref
+                HeadBranch = pr.Head.Ref,
+                Reviewers = reviewers
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs]: Add a triailing comma to avoid this kind of diff in the future. :-)

**Comment by AdamYoblick** [comment]
Thanks for all the feedback guys. This got put on the back burner for me, then I got sick, and I’m taking pat leave in less than 3 weeks, so this is going to sit for a while :)

________________________________

**Comment by markwilkie** [comment]
Congrats @adamyoblick

**Comment by mmitche** [comment]
> Congrats @AdamYoblick

At first I thought you were congratulating @AdamYoblick on getting a proper @ChadNedzlek code review.

**Comment by AdamYoblick** [comment]
Haha! :)

________________________________

**Comment by markwilkie** [comment]
lol

File: src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs (hunk 4 above)
Comment (ChadNedzlek) [reply src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs]: Well we did create the review on line 299, but the caller is going to, presumably, retry (internet and such) not sure that that will do.

---
In reply to: [254760371](https://github.com/dotnet/arcade-services/pull/88#discussion_r254760371) [](ancestors = 254760371)

**Comment by mmitche** [comment]
I'm closing this as it is really out of date

**Pull request closed without merging**